staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go

```diff
@@ -35,19 +35,17 @@ const (
 	StatusTooManyRequests = 429
 )
 
-// StatusError is an error intended for consumption by a REST API server; it can also be
-// reconstructed by clients from a REST response. Public to allow easy type switches.
-type StatusError struct {
-	ErrStatus metav1.Status
-}
-
 // APIStatus is exposed by errors that can be converted to an api.Status object
 // for finer grained details.
 type APIStatus interface {
-	Status() metav1.Status
+	Status() *metav1.Status
```

Comment:
the original intent of this interface was to allow multiple types of errors to provide status, not depend on a single error implementation (https://github.com/kubernetes/kubernetes/commit/ff2eca97d9f202fb05f35582f2dad57fd53a1450#diff-fb87476fa66849c8514ba700e715b4b6R85)

I don't think the intent was necessarily to return a runtime object

Returning a pointer raises the possibility of returning nil, and raises the question of needing to check for nil at callsites.

Reply:
> the original intent of this interface was to allow multiple types of errors to provide status, not depend on a single error implementation (ff2eca9#diff-fb87476fa66849c8514ba700e715b4b6R85)

that intent is being kept

> I don't think the intent was necessarily to return a runtime object

My reasoning about it is - if you return an object which has pointer receiver methods you should return it as pointer otherwise that object is hard to use.

so instead of e.g.
```golang
watch.Event{
	Type:   watch.Error,
	Object: apierrors.NewInternalError(someError).Status(),
}
```

it forces people to manually fix the bad interface by a wrapper:

```golang
func apiStatusPtr(status metav1.Status) *metav1.Status {
    return &status
}

watch.Event{
	Type:   watch.Error,
	Object: apiStatusPtr(apierrors.NewInternalError(someError).Status()),
}
```

which seems silly and like a bug of the interface design.

`metav1.Status` is described as "[Status is a return value for calls that don't return other objects.](https://github.com/kubernetes/kubernetes/blob/1cc316f4dc7c576c6fd0fe6701671fc530d9bf28/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L522)" which implies such object is supposed to be handled like any other `runtime.Object` which is always passed by pointer in kube.

> Returning a pointer raises the possibility of returning nil, and raises the question of needing to check for nil at callsites.

Seems like a general language issue we have with any other object that's being passed by pointer even though it's not optional.

Also with the value semantics you are making a false promise of a copy, by actually doing just a shallow copy, which is bug prone. Returning a pointer would make it clear this is not a copy.

Reply:
> if you return an object which has pointer receiver methods you should return it as pointer otherwise that object is hard to use.

in non-test code, this PR makes the returned object harder to use in three places (`ErrStatus: *s.Status()`), and easier in two (`ErrorToAPIStatus` and `transformErrorToEvent`).

> Also with the value semantics you are making a false promise of a copy, by actually doing just a shallow copy, which is bug prone. Returning a pointer would make it clear this is not a copy.

The shallow copy was sufficient for the modifications made in `ErrorToAPIStatus`. To be equivalent to the behavior prior to this PR, `ErrorToAPIStatus` would need to go back to making a shallow copy, which means this PR doesn't actually improve usage for that call site.

I don't see a strong reason to make this change.

Reply:
> in non-test code, this PR makes the returned object harder to use in three places (ErrStatus: *s.Status()), and easier in two (ErrorToAPIStatus and transformErrorToEvent).

hmm, so statistics :) if the interface makes the other uses cumbersome or syntactically unusable does the current usage counts?  I have a PR (https://github.com/kubernetes/kubernetes/pull/67350) in flight where this makes 9 new usages possible and simple, turning the odds